Machine handler and service layer

GHFluding opened the pull request.

GHFluding committed:
Machine handler and service layer

GHFluding committed:
added all services and models

GHFluding committed:
Fixes

Added function to run on import(App.RunExternal). Fixed name inaccuracies. Added structures to make the interface concise

GHFluding committed:
Update main.go

tritonJS826 commented on SMgrpc/internal/domain/models/shift.go:
```diff
@@ -1,6 +1,6 @@
-	MachineId int64
-	MasterId  int64
+	MachineId   int64
+	ShiftMaster int64
```
ShiftMasterId?

tritonJS826 commented on SMgrpc/internal/grpc/sm/shift/server.go:
```diff
@@ -30,15 +33,15 @@ func RegisterServerAPI(gRPC *grpc.Server, shift ShiftInterface) {
-			MachineId:   machine,
-			ShiftMaster: master,
+			MachineId:   shift.MachineId,
+			ShiftMaster: shift.ShiftMaster,
```
ShiftMasterId?

tritonJS826 commented on SMgrpc/internal/grpc/sm/task/server.go:
```diff
@@ -9,6 +9,13 @@ import (
+	MachineId    int64
+	ShiftId      int64
+	Frequency    string
+	TaskPriority string
```
is it really string? Or should it be enum?

tritonJS826 commented on SMgrpc/internal/app/app.go:
```diff
@@ -2,17 +2,52 @@ package app
+	application := New(log, cfg.GRPC.Port, db_app, mode)
+
+	//maybe use run
+	go application.GRPCServer.MustRun()
```
nice comment :)

GHFluding committed:
fixes and comments

GHFluding committed:
added go work and renamed modules

GHFluding committed:
fixed misstackes

GHFluding committed:
added last machine interface

imported parts moved to pkg from internal

tritonJS826 commented on SM/configs/sqlc/machine_query.sql:
```diff
@@ -18,4 +18,6 @@ WHERE id = @id;
-
+-- name: GetMachine :one
+SELECT * FROM Machine
+WHERE id = $1;
```
you can use named parameters: `$1` -> `@machineId`. It will help you to male query and query object itself  more readable for people!

GHFluding committed:
fixed misstakes

GHFluding committed:
added user db function for grpc interface

tritonJS826 commented on SM/internal/grpcStorage/user/user.go:
```diff
@@ -0,0 +1,84 @@
+	if err != nil {
+		return postgres.CreateUserParams{}, err
+	}
+	var bitrixidDB pgtype.Int8
+	if bitrixId == nil {
+		bitrixidDB.Valid = false
+	} else {
+		bitrixidDB.Valid = true
+		bitrixidDB.Int64 = *bitrixId
+	}
```
it will be good to create helper-generic for such *int -> pgtype.Int

GHFluding committed:
added convertor and interface function for task and shift

GHFluding committed:
grpc server and client

Added grpc client and server. Added a file to run the application with grpc. Fixed errors with interfaces in the part of server and DB communication.

tritonJS826 commented on SMgrpc/pkg/client/machine.go:
```diff
@@ -0,0 +1,21 @@
+)
+
+func (c *Client) CreateMachine(ctx context.Context, name string, isRepairRequired, isActive *bool) (*entities.MachineResponse, error) {
+	ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
```
magic number 3 - provide a name for this value and move to shareable place

tritonJS826 commented on SMgrpc/pkg/client/shift.go:
```diff
@@ -0,0 +1,20 @@
+	entities "github.com/GHFluding/ShiftManager/SMgrpc/internal/gen"
+)
+
+func (c *Client) CreateShift(ctx context.Context, machineID, shiftMaster int64) (*entities.ShiftResponse, error) {
```
shiftMaster -> shiftMasterId ?

GHFluding committed:
fix

GHFluding committed:
added grpc service functions for link service

tritonJS826 commented on link/internal/services/machine.go:
```diff
@@ -66,6 +68,35 @@ func CreateMachine(data []byte, log *slog.Logger, url string) ([]byte, error) {
 	return responseData, nil
 }
 
+func CreateMachineGRPC(c *client.Client, data []byte, log *slog.Logger, url string) ([]byte, error) {
```
why do yo have data with []byte type? If it is a grpc server it should be possible to generate internal models. The same for rest service. 
[]byte - is a generic, but your schema could be defined in the code

GHFluding committed:
fixed response

GHFluding committed:
added validator

GHFluding committed:
fix

GHFluding committed:
added for all

tritonJS826 commented on link/internal/services/machine.go:
```diff
@@ -69,21 +69,17 @@ func CreateMachine(data []byte, log *slog.Logger, url string) ([]byte, error) {
 	}
 
-	resp, err := c.CreateMachine(context.Background(), machine.Name, machine.IsRepairRequired, machine.IsActive)
+	resp, err := c.CreateMachine(context.Background(), data.Name, data.IsRepairRequired, data.IsActive)
```
Объединяй в структуры, у объектов и по 20 полейц бывает) Почему не принять просто *entities.CreateMachine? - так и писать меньше будет и с расширениями в будущем на запутаешся

GHFluding committed:
webhook for grpc

Made a webhook for grpc, also some minor fixes

GHFluding committed:
fix

tritonJS826 commented on link/internal/services/machine.go:
```diff
@@ -22,7 +22,7 @@ type createMachineParams struct {
 }
 
-func CreateMachine(data []byte, log *slog.Logger) ([]byte, error) {
+func CreateMachine(data []byte, log *slog.Logger, url string) ([]byte, error) {
```
again - []byte params looks wrong a little

check twice is url required - why should it be param, not hardcoded constant.
